crates/signature-mpc/src/twopc_mpc_protocols/dwallet_transfer.rs

```diff
@@ -0,0 +1,361 @@
+    plaintext: Vec<u8>,
+    language_public_parameters: encryption_of_discrete_log::PublicParameters<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        SCALAR_LIMBS,
+        twopc_mpc::secp256k1::GroupElement,
+        EncryptionKey,
+    >,
+) -> ProofPublicOutput {
+    let paillier_public_parameters: tiresias::encryption_key::PublicParameters =
+        bcs::from_bytes(&encryption_key).unwrap();
+
+    let unbounded_witness_public_parameters = language_public_parameters
+        .randomness_space_public_parameters()
+        .clone();
+
+    let plaintext = parse_plaintext(plaintext, &paillier_public_parameters);
+
+    let randomness = RandomnessSpaceGroupElement::sample(
+        language_public_parameters
+            .encryption_scheme_public_parameters
+            .randomness_space_public_parameters(),
+        &mut OsRng,
+    )
+    .unwrap();
+
+    let enhanced_language_public_parameters = enhanced_language_public_parameters::<
+        { maurer::SOUND_PROOFS_REPETITIONS },
+        RANGE_CLAIMS_PER_SCALAR,
+        RandomnessSpaceGroupElement,
+        Lang,
+    >(
+        unbounded_witness_public_parameters,
+        language_public_parameters,
+    );
+
+    let witnesses = generate_witnesses(randomness, &enhanced_language_public_parameters, plaintext);
+
+    let (proofs, statements) = SecretShareProof::prove(
+        &PhantomData,
+        &enhanced_language_public_parameters,
+        witnesses,
+        &mut OsRng,
+    )
+    .unwrap();
+
+    ProofPublicOutput {
+        proof: proofs,
+        encrypted_discrete_log: statements[0]
+            .language_statement()
+            .encrypted_discrete_log()
+            .value(),
+        range_proof_commitment: statements[0].range_proof_commitment().value(),
+    }
+}
+
+fn generate_witnesses(
+    randomness: RandomnessSpaceGroupElement,
+    enhanced_language_public_parameters: &language::PublicParameters<
+        { maurer::SOUND_PROOFS_REPETITIONS },
+        EnhancedLang<
+            { maurer::SOUND_PROOFS_REPETITIONS },
+            RANGE_CLAIMS_PER_SCALAR,
+            RandomnessSpaceGroupElement,
+            Lang,
+        >,
+    >,
+    plaintext: PlaintextSpaceGroupElement,
+) -> Vec<
+    WitnessSpaceGroupElement<
+        { maurer::SOUND_PROOFS_REPETITIONS },
+        RANGE_CLAIMS_PER_SCALAR,
+        COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS,
+        bulletproofs::RangeProof,
+        RandomnessSpaceGroupElement,
+        Lang,
+    >,
+> {
+    let witnesses: Vec<language::WitnessSpaceGroupElement<1, Lang>> =
+        vec![(plaintext, randomness).into()];
+
+    let witnesses =
+        EnhancedLanguage::<
+            { maurer::SOUND_PROOFS_REPETITIONS },
+            RANGE_CLAIMS_PER_SCALAR,
+            { COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS },
+            bulletproofs::RangeProof,
+            RandomnessSpaceGroupElement,
+            Lang,
+        >::generate_witnesses(witnesses, &enhanced_language_public_parameters, &mut OsRng)
```

Comment:
Use `generate_witness`

Reply:
Wdym? Aren't we already using `generate_witness`?